dask_expr/reductions.py

```diff
@@ -151,26 +267,86 @@ def aggregate(cls, inputs, **kwargs):
         df = _concat(inputs)
         return func(df, **kwargs)
 
-    def __dask_postcompute__(self):
-        return toolz.first, ()
 
-    def _divisions(self):
-        return [None, None]
+##
+## Templates
+##
+
+
+class MapReduceTemplate:
+    """MapReduce sub-class template
+
+    This non-`Expr` class defines attributs that are
+    shared between the 'map' and 'reduce' stages of a
+    typical map-reduce operation. The purpose of this
+    class is to reduce the amount of boiler-plate code
+    needed to define a `MapReduce` sub-class.
+
+    Typical usage:
+
+        # Define a template
+        class MyTemplate(MapReduce):
+            # Define common attributes
+            _parameters = [...]
+            ...
+
+        # Use the template to define an expression
+        class MyExpr(MyTemplate, MapReduce):
+            # Generate `map` and `reduce`
+            map = MyTemplate.make_map()
+            reduce = MyTemplate.make_reduce()
```

Comment:
A key question for this PR is whether we *want* something like this "template" idea to make it possible to reduce the amount of code needed to implement a specific reduction. For example, without the template, `Sum` would need to look something like this:

```python
class SumMap(DataFrameMap):
    _parameters = ["frame", "skipna", "numeric_only", "min_count"]
    _defaults = {"skipna": True, "numeric_only": None, "min_count": 0}
    reduction_chunk = M.sum

    @property
    def chunk_kwargs(self):
        return dict(
            skipna=self.skipna,
            numeric_only=self.numeric_only,
            min_count=self.min_count,
        )

class SumReduce(DataFrameReduce):
    _parameters = ["frame", "skipna", "numeric_only", "min_count"]
    _defaults = {"skipna": True, "numeric_only": None, "min_count": 0}
    reduction_aggregate = M.sum

class Sum(DataFrameReduction):
    map = SumMap
    reduce = SumReduce

    def _simplify_up(self, parent):
        if isinstance(parent, Projection):
            return self.frame[parent.operand("columns")].sum(*self.operands[1:])
``` 

Which is only a bit more code than we need with the "template" mechanism:

```python
class _Sum(DataFrameReductionTemplate):
    _parameters = ["frame", "skipna", "numeric_only", "min_count"]
    _defaults = {"skipna": True, "numeric_only": None, "min_count": 0}
    reduction_chunk = M.sum
    reduction_aggregate = M.sum

    @property
    def chunk_kwargs(self):
        return dict(
            skipna=self.skipna,
            numeric_only=self.numeric_only,
            min_count=self.min_count,
        )

    def _simplify_up(self, parent):
        if isinstance(parent, Projection):
            return self.frame[parent.operand("columns")].sum(*self.operands[1:])


class Sum(_Sum, DataFrameReduction):
    map = _Sum.make_map()
    reduce = _Sum.make_reduce()
```

Reply:
Given this comparison, my current preference is to simply define each `Map`/`Reduce` component explicitly (and remove the over-engineered template mechanism).

Reply:
I could also be wrong about this.  Maybe lots of people use shuffle groupbys for example.  I don't have evidence of that though.  I'm open to being educated otherwise.

Reply:
An alternative implementation (which I'd suggest we delay) is that we create a ShuffleApplyConcatApply expression like the following:

```python
class ApplyConcatApply(Expr):
    ...

class ShuffleApplyConcatApply(ApplyConcatApply):
    def _layer(self):  # the only thing we need to add
```

And then we subclass all of the reductions (I would hope that we find something better) but with nothing extra (or maybe an extra split_out parameter, but that's all)

```python
class ShuffleSum(Sum, ShuffleApplyConcatApply):
    pass
```

I don't particularly like this either, but it does show us that we can add shuffling in in the future while still keeping the current Sum/Max/Min reductions easy to write and understand.  The thing that changes is `_layer`.  That's all that we should really care about I think.

Maybe we can even automate the creation of the `SumShuffle` expressions by generating them on demand.

In general, I find that I'm reacting strongly to anything that makes writing the Sum/Max/Min logic more complicated.  I think that we need it to be very easy to author and maintain algorithms.

Reply:
I don't see us needing to support a shuffle-based aggregation for most reductions. Just operations like groupby and unique/drop_duplicates.

In the long run, I suppose I'm just fairly confident that we will want to be working with abstract ACA/Reduction expressions that can be converted into something like `chunk` + `shuffle` + `aggregate` or `chunk` + `repartition` + `aggregate` or just `chunk` + `aggregate`, depending on top-level arguments (like `split_out` etc..).

I'm not a fan of the current approach in this PR - It's just an experiment. I much preferred the simple approach of splitting something like into chunkwise and reduction phases by passing in the necessary chunk/combine/aggregate components as operands. That approach did not require any changes to the existing reductions, but resulted in hard-to-read `pprint` results after `simplify` was called.

Reply:
> Just operations like groupby and unique/drop_duplicates

In this case then maybe the `class ShuffleSum(Sum, ShuffleApplyConcatApply)` approach works well.  We make `DropDuplicates(ApplyConcatApply)` and then make a variant algorithm that inherits from a shuffle algorithm.

```python
class DropDuplicatesShuffle(ShuffleACA, DropDuplicates):
    pass
```